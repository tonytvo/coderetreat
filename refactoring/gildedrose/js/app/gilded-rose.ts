import {QualityUpdater} from "@/quality-updater";
import {BackstagePassQualityUpdater} from "@/backstage-pass-quality-updater";
import {AgedBrieQualityUpdater} from "@/aged-brie-quality-updater";
import {SulfurasQualityUpdater} from "@/sulfuras-quality-updater";
import {DefaultItemQualityUpdater} from "@/default-item-quality-updater";

export class NonNegativeWithCeilingQuality {
    quality: number;
    constructor(quality) {
        this.quality = quality;
    }

}

export class Item {
    name: string;
    sellIn: number;
    quality: number;
    qualityTemp: NonNegativeWithCeilingQuality;

    constructor(name, sellIn, quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
        this.qualityTemp = new NonNegativeWithCeilingQuality(quality);
    }

    public toString() {
        return `{"name":"${this.name}","sellIn":${this.sellIn},"quality":${this.quality}}`;
    }
}

export class GildedRose {
    items: Array<Item>;

    constructor(items = [] as Array<Item>) {
        this.items = items;
    }

    updateQuality() {
        for (let i = 0; i < this.items.length; i++) {
            let item = this.items[i];
            this.createQualityUpdater(item.name).update(item);
        }

        return this.items;
    }

    private createQualityUpdater(name: string) : QualityUpdater{
        if (name == 'Aged Brie') {
            return new AgedBrieQualityUpdater();
        }

        if (name == 'Backstage passes to a TAFKAL80ETC concert') {
            return new BackstagePassQualityUpdater();
        }

        if (name == 'Sulfuras, Hand of Ragnaros') {
            return new SulfurasQualityUpdater();
        }

        return new DefaultItemQualityUpdater();
    }
}
