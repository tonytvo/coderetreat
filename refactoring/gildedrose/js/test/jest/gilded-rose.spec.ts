import {GildedRose, Item} from '@/gilded-rose';
import 'jest-extended-snapshot';

describe('Gilded Rose', () => {
  function updateGildedRoseQuality(name: string, quality: number) {
    const gildedRose = new GildedRose([new Item(name, 0, quality)]);
    const items = gildedRose.updateQuality();
    return JSON.stringify(items[0]);
  }

  it('should foo', () => {
    let names = ['foo'
      ,"+5 Dexterity Vest"
      ,"Aged Brie"
      ,"Elixir of the Mongoose"
      ,"Sulfuras, Hand of Ragnaros"
      ,"Backstage passes to a TAFKAL80ETC concert"];
    let qualities = [-1, 0, 1, 49, 50, 51];
    expect(updateGildedRoseQuality).toVerifyAllCombinations(names, qualities);
  });
});
