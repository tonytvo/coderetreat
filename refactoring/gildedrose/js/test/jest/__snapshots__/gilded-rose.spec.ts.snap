// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose should foo 1`] = `
Object {
  "+5 Dexterity Vest,-1": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":-1}",
  "+5 Dexterity Vest,0": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "+5 Dexterity Vest,1": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "+5 Dexterity Vest,49": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":47}",
  "+5 Dexterity Vest,50": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":48}",
  "+5 Dexterity Vest,51": "{\\"name\\":\\"+5 Dexterity Vest\\",\\"sellIn\\":-1,\\"quality\\":49}",
  "Aged Brie,-1": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":1}",
  "Aged Brie,0": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":2}",
  "Aged Brie,1": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":3}",
  "Aged Brie,49": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":50}",
  "Aged Brie,50": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":50}",
  "Aged Brie,51": "{\\"name\\":\\"Aged Brie\\",\\"sellIn\\":-1,\\"quality\\":51}",
  "Backstage passes to a TAFKAL80ETC concert,-1": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Backstage passes to a TAFKAL80ETC concert,0": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Backstage passes to a TAFKAL80ETC concert,1": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Backstage passes to a TAFKAL80ETC concert,49": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Backstage passes to a TAFKAL80ETC concert,50": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Backstage passes to a TAFKAL80ETC concert,51": "{\\"name\\":\\"Backstage passes to a TAFKAL80ETC concert\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Elixir of the Mongoose,-1": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":-1}",
  "Elixir of the Mongoose,0": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Elixir of the Mongoose,1": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "Elixir of the Mongoose,49": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":47}",
  "Elixir of the Mongoose,50": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":48}",
  "Elixir of the Mongoose,51": "{\\"name\\":\\"Elixir of the Mongoose\\",\\"sellIn\\":-1,\\"quality\\":49}",
  "Sulfuras, Hand of Ragnaros,-1": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":-1}",
  "Sulfuras, Hand of Ragnaros,0": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":0}",
  "Sulfuras, Hand of Ragnaros,1": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":1}",
  "Sulfuras, Hand of Ragnaros,49": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":49}",
  "Sulfuras, Hand of Ragnaros,50": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":50}",
  "Sulfuras, Hand of Ragnaros,51": "{\\"name\\":\\"Sulfuras, Hand of Ragnaros\\",\\"sellIn\\":0,\\"quality\\":51}",
  "foo,-1": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":-1}",
  "foo,0": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "foo,1": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":0}",
  "foo,49": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":47}",
  "foo,50": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":48}",
  "foo,51": "{\\"name\\":\\"foo\\",\\"sellIn\\":-1,\\"quality\\":49}",
}
`;
